### DISCUSSION TITLE
Migrating to 2023 PathPlanner

### DISCUSSION THREAD

**Discussion by gcschmit** [mjansen4857/pathplanner#158] [Q&A] [answered]
@mjansen4857 Thank you for providing the FRC community with an incredible tool. Our team found PathPlanner very useful last season. I'm impressed with the new features that you have made for the 2023 season!

Our team is currently evaluating the 2023 wpilib beta by updating last year's code to the 2023 wpilib beta and latest versions of third-party tools like PathPlanner. One change that we noticed was that the PPSwerveControllerCommand now takes a PIDController instead of a ProfiledPIDController for the rotation controller. I noticed that the PPHolonomicDriveController class invokes the enableContinuousInput on the rotation controller; so, we no longer need to do so.

More generally, should we be concerned about losing the ability to specify the maximum rotational acceleration and velocity via TrapezoidProfile.Constraints like we could do last year with the ProfiledPIDController? Is that now handled elsewhere? Or, do we need to ensure we handle it elsewhere?

Thanks again for all your efforts to provide teams with PathPlanner!

**Comment by mjansen4857** [comment, marked as answer]
The reason for this change is because PPHolonomicDriveController uses holonomic angular velocity, and then applies correction from a normal PID controller based on the rotational error. This is how the X/Y controllers work already. You will lose the ability to have the constraints of a profiled PID controller, so you will need to take care when creating paths to make sure that what you're telling the robot to do is something it can actually achieve. For example, trying to do 180 while going full speed over a short distance just wont happen. But this is the same as the old version, the outcome will just be a bit different. Either way you'd have some inaccuracy from the rotation and translation fighting over saturation of the module if that makes sense.

In the future I want the generation to take rotation velocity into account when determining translation velocity to prevent this. Not doable for this year though since the generation would need a total-rework for stuff like that to work.

**Reply by gcschmit** [reply]
That does make sense. We will take care when creating our paths. Thanks!

**Comment by WarrenReynolds** [comment]
We built our follower from modified WPILib sample code so I can see where you are coming from this post.

I can see a future version of PPHolonomicDriveController having a feature where the holonomic heading is either determined by values in  a PPtrajectory, in this case a Normal PID Controller for rotation is fine as the trajectory generator effectively smooths out the rotation as it travels along the path.

But I can also see a future option being added where the holonomic rotation is optionally controlled by an external lambda function for certain use cases. (eg on button press) An example of this is some code that always holds the robot facing a set direction or actively tracking a target while an on-the-fly trajectory has been generated and is being followed to return to a loading bay.  In this case, a ProfiledPID rotational controller would be a nicer option for the Holonomic heading Controller as the robot might be completely facing the wrong way when this feature is activated, which would cause a very large rotation rate correction being applied to the holonomic heading because of the Kp on the holonomic error.

So I can see @gcschmit point were a profiled PID controller would handle this situation better.  However it does mean more parameters need to be passed into the controller.

To give a full example case from this years game.  A ball has just been picked up from the corner of the field. The robot is facing directly away from the goal.  An on-the-fly PPpath is generated to take the robot from where it is, to the "sweet spot" out in the open where every shot goes in.  During this PPtrajectory following, the holonomic heading could be controlled from external code that aligned the robot with the target for the entire path, that way if the drivers saw that they weren't going to reach the sweet spot they could fire anytime before hand and the robot would be facing the correct direction.  The profiled controller would smoothly spin the robot to face the target.

It really only needs to be changed to a Profiled PID controller if the ability to set the holonomic heading while following a path is optionally handed over to an external source from inside the follower.  

We haven't used PP on-the-fly generation yet, but we did have buttons that disconnected the z axis on the joystick and forced the robot to always face the target (With a profiled PID rotational controller), so I can see us hopefully combining these features for next year.  Also setting it to a profiled PID Controller would align it with WPILib's approach to holonomic heading control.  Its a great product either way.

**Reply by mjansen4857** [reply]
I've actually planned on looking into something like this before kickoff. The ability to aim while driving is a big advantage for swerve so trying to support that is pretty important for me.

**Reply by WarrenReynolds** [reply]
Love your work.

**Reply by mjansen4857** [reply]
Actually, thinking about this a bit more, I believe this can be accomplished in a cleaner way by just outputting field relative chassis speeds from the path following command as has already been suggested. Then, you can choose to either pass this chassis speeds object to your kinematics to get your module states, or edit the angular velocity of the chassis speeds before doing so to the output of a pid controller running on vision data for example.

Doing it this way would avoid the ugly need to pass around some "override" value to the path following commands.

**Reply by WarrenReynolds** [reply]
Is the follower still going to require the drive subsystem?

**Reply by mjansen4857** [reply]
Yeah. It would just be a different output option. You can choose to output in module states or ChassisSpeeds. Either way its still "controlling" the drive subsystem with this output. It's just one is more customizable. You technically could do the same thing already in your outputModuleStates method to convert your module states to ChassisSpeeds with kinematics, change the rotation velocity, then convert back to module states. That is just obviously worse to work with.

**Reply by WarrenReynolds** [reply]
My head is starting to melt again. Help.
So when you create a PPSwerveControllerCommand you are planning to supply a consumer that takes a ChassisSpeed object instead of passing in a consumer that takes the swerve module states and then in the drive subsystem periodic code you can optionally take over the heading control of the chassisSpeed object before sending the module states to swerve modules.

I'd like to be able to select who is in control of the holonomic heading from Marker or Stop point commands that are executed along the paths.  So these commands would set flags in the drive subsystem to set where the heading control is coming from.  Is this right?

I think this is aligned with what we did this year with our activetrack and heading lock code. So I think I like this.

**Reply by WarrenReynolds** [reply]
I'm pretty keen to implement this for our team for the upcoming season. Can I ask your opinion on these modifications to implement.

In PPSwerveControllerCommand
Rename m_output member variable to m_outputModuleStates
Create a new member variable called m_outputChassisSpeeds
At the end of PPServeConrollerCommand:Execute
Change 

`this->m_output(targetModuleStates);`

to 

`this->m_outputModuleStates(targetModuleStates);`

add

`this->m_outputChassisSpeeds(targetChassisSpeeds);`

In regards to the constructors, do you 

Create a new constructor that has everything the old constructor had plus an extra parameter to initialize m_outputChassisSpeeds.
Modify all the exiting constructors so they initialize using the above new constructor with a "do nothing lambda" consumer for outputChassisSpeeds  (like you did with the resetPose), to maintain backward compatibility.

Duplicate all existing constructors, using the same principle as above but with a "do nothing Lambda" consumer on the TargetModuleStates.  So these ones we will use the chassisSpeeds that are returned from PPSwerveControllerCommand instead of the module states.

or could you just make two option parameters that default to "do nothing lambda" for the output consumers? Then depending on what arguments are supplied to the constructor of PPSwerveControllerCommand (either a consumer that takes Chassis Speeds or a consumer that takes Module States) will depend on what output you use. NB: I can see an issue that a PPSwerveControllerCommand could be constructed that doesn't do anything if both optional parameters are omitted, so I assume this is not considered best practice.

In all honesty, If we are going to make our own custom version there is no need to maintain all the constructors that are never going to be used.  I was just thinking something like this so that when you release the next version in 2023 that has this feature I'm hoping it will drop straight in.

Then create an enumeration in constants.h for HolonomicControl with values of say PathPlanner, Joystick, LockedHeading, TargetLockOdometry, TagetLockLimelight, FaceHordCorner JingDefenceAvoidance, plus however else you want to control the holonomic heading.

In Drive Subsystem create a private member value that holds a HolonomicConrol Enum, create a public setter and getter for this member variable that commands can use to set, then in your setModuleStatesFromChassisSpeeds method of the drive subsystem, query this member variable to see where the holonomic rotation should come from and control accordingly.

Then have the default drive command act through calls to setModuleStatesFromChassisSpeeds so when nothing has the drive subsystem required the joystick is controlling the drive.

Thoughts?

**Reply by gcschmit** [reply]
> Yeah. It would just be a different output option. You can choose to output in module states or ChassisSpeeds. Either way its still "controlling" the drive subsystem with this output. It's just one is more customizable. You technically could do the same thing already in your outputModuleStates method to convert your module states to ChassisSpeeds with kinematics, change the rotation velocity, then convert back to module states. That is just obviously worse to work with.

What I really appreciate about PathPlanner is that it the team's code is in control at the top-level (e.g., when creating commands that subclass PPSwerveControllerCommand and in control at the button-level (e.g., when receiving swerve module states). Even if it may be inconvenient, because of this design most anything is possible. I think a strength of PathPlanner is that it makes the common cases so much easier for teams and yet still makes to odd or advanced cases possible.

**Reply by gcschmit** [reply]
@WarrenReynolds I may not fully understand your use case, but I believe you can remain in total control of the holonomic rotation without any changes to PathPlanner. In your drive subsystem when it receives the swerve module states from PathPlanner, you can covert those back to chassis speeds:

```
chassisSpeeds = kinematics.toChassisSpeeds(swerveModuleStates[0], ...);
chassisSpeeds.omegaRadiansPerSecond = <whatever you want>;
swerveModuleStates = kinematics.toSwerveModuleStates(chassisSpeeds);
```

**Reply by mjansen4857** [reply]
This is what i meant when I said:

> You technically could do the same thing already in your outputModuleStates method to convert your module states to ChassisSpeeds with kinematics, change the rotation velocity, then convert back to module states. That is just obviously worse to work with.

The new output option for Chassis speeds would just get rid of that first conversion from module states.
